### PR TITLE
Implement implicit feedback transformer and measurement for top-N.

### PR DESCRIPTION
**Pull request by YuyangZhangFTD**
Add implicit feedback transformer in Reader class. 
Add hr and arhr accuracy measurement for implicit feedback in accuracy module. 
Fix bug with fcp in accuracy module.

### CODE REVIEW

**Comment by YuyangZhangFTD** [comment]
@NicolasHug I suggest we should remove python2 in travis-ci. 
Numpy won't support python2 in the future, so...

**Comment by NicolasHug** [comment]
Hi, thanks a lot for this.

As you noted HR and ARHR were introduced in a LOO setting: the testset is made of one positive interaction for each user (so |testset| = |U|). Unfortunately I'm not sure the way you're dealing with it right now really is compatible with this view.
This is something that I already had in mind: if we are to implement HR and ARHR, I think that we would need to upgrade the cross-validation tools and propose LOO or more generally LPO.

That being said, allowing to binarize an explicit dataset to convert it into an implicit one is a good idea and a nice feature to have. That would require extensive testing though: I fear that as it's currently implemented, there would be some clash with the `offset` and `rating_scale` attributes of the `Reader` object.

Nicolas

**Comment by YuyangZhangFTD** [comment]
Hey,
I'm not very clear with LOO and LPO...Could you explain it? Plz...
And you are right about 'offset'. I wil make it compatible with 'offset' soon.

Yuyang Zhang

**Comment by NicolasHug** [comment]
Leave-One-Out is exactly what is explained in the paper you cited: for each user, we hide one positive interaction during the training and put all those n = |U| interactions in the testset.

Leave-P-Out is the same except that we hide P interactions instead of one.

**Comment by YuyangZhangFTD** [comment]
Thanks for explaining~
So,  is whether the testset contains all users what you worry about ?

**Comment by YuyangZhangFTD** [comment]
Thanks for explaining.
In HR and ARHR, some test samples are abandoned for there are not enough test samples for top-n list or no positive samples.  When I test in ml-100k, about 1/4 users are abandoned in 5-fold test set.

**Comment by NicolasHug** [comment]
Yes, that's one of my concerns. In the current implementation, you put all predictions with a ground truth rating of 1 in the `clicked_list` and all the others in the `testset`? But as the dataset is comprised of only 1 ratings, everything is in `clicked_list` and only one random rating ends up in `testset`. This is not how those metrics were introduced in the papers.

**Comment by YuyangZhangFTD** [comment]
@NicolasHug I've worked on hr and arhr for a few days and I have 2 plans:
1. Try to re-construct trainset and testset for implicit feedback, in which, we're going to implement HR and ARHR just as what they did in those papers. It may take a lot of time.
2. Just abandon this feature this time, and make a future plan. Maybe hr and arhr are not suitable for Surprise. We can implement other measurements for implicit feedback.

**Comment by NicolasHug** [comment]
Yes this is all a bit complicated. My take on the matter is the following:

Before implementing HR and ARHR or any other accuracy metric, we need to be able to [test](https://github.com/NicolasHug/Surprise/blob/master/.github/CONTRIBUTING.md#contributing-to-surprise) them.
To test them, we need an implicit feedback prediction algorithm, which should as well be tested.
And to build an implicit feedback prediction algorithm, we need a way to handle implicit feedback. The way you proposed (binarizing an explicit dataset) is one possible way of doing that. But it still has to be tested!

Additionally, RH and ARHR would need us to implement some extra cross-validation procedures. This should not be too hard but it needs to be done carefully and we'll need to decide whether we'll break the current API or not. Ideally, I'd like to have something as close as possible to scikit-learn's API, which would require some refactoring and some API breaking. Breaking the API should be avoided as much as possible but I fear that allowing to manage implicit feedback **will** require us to change it eventually.

So, I believe we first need to start with little easy things, and the iterate from that. A possible course of action could be:
- implement dataset binarization (done but not tested yet. To test it, we probably need the second step:)
- implement a very basic implicit prediction algorithm
- implement basic accuracy metrics. Precision and Recall should be fairly easy, and ideally they should be also usable in the explicit case.

Obviously each of those steps should be thoroughly tested.

Then, when the basic blocks are working, we can start more fancy stuff like better prediction algorithms (BPR and SLIM are quite simple I think), other ways to manage implicit dataset, ARHR and HR, etc...

I know this is **a lot** of work, so no worries if you feel like this is too much. In any case if you decide to give it a try, we would need to discuss it a lot as there would be many choices to be made.

Nicolas

**Comment by NicolasHug** [comment]
Hey, coming back to you on this: I'll soon start to work on implicit feedback algorithms and dataset binarization. Do you mind if I close this PR?

**Comment by YuyangZhangFTD** [comment]
Hey, I'm so sorry for this. I'm busy doing with my work these days, so...
Maybe you can do this better than me. I will continue focusing on it, and do what I can do for surprise.
BTW, surprise is so awesome : )

**Comment by NicolasHug** [comment]
No worries at all, this is a fairly tricky problem, and especially hard for a first contribution.
Thanks a lot for the quick feedback and the kind words!
